kimchi/src/circuits/expr.rs

```diff
@@ -3149,3 +3149,246 @@ pub mod test {
+        Col: ColTrait + Hash + Eq,
+    {
+        /// returns self transformed into a degree 2 or less constraint and a set of additional constraints,
+        /// each for one of the new columns
+        pub fn reduce_degree_to_2(
+            self,
+        ) -> (
+            Expr<C, ColumnExtended<Col>>,
+            Vec<Expr<C, ColumnExtended<Col>>>,
+        ) {
+            let mut rec = ExprRecorder::new();
+            let exp = self.reduce_degree_to_2_rec(&mut rec);
+            let new_constraints = rec.into_constraints();
+            (exp, new_constraints)
+        }
+        fn pow_to_mul(self, p: u64) -> Self {
+            let e = Box::new(self);
+            let mul = |a, b| Expr::BinOp(Op2::Mul, a, b);
+            let e_2 = Box::new(Expr::Square(e.clone()));
+            let new_exp = match p {
+                2 => *e_2,
+                3 => mul(e, e_2),
+                4..=8 => {
+                    let e_4 = Box::new(Expr::Square(e_2.clone()));
+                    match p {
+                        4 => *e_4,
+                        5 => mul(e, e_4),
+                        6 => mul(e_2, e_4),
+                        7 => mul(e, Box::new(mul(e_2, e_4))),
+                        8 => e_4.square(),
+                        _ => unreachable!(),
+                    }
+                }
+                _ => panic!("unsupported"),
+            };
+            new_exp
+        }
+        fn reduce_degree_to_2_rec(
+            self,
+            rec: &mut ExprRecorder<C, Col>,
+        ) -> Expr<C, ColumnExtended<Col>> {
+            let trivial_map = |c| ColumnExtended::Inner(c);
+            let degree = self.degree(4) / 4;
+            if degree <= 2 {
+                return self.map_column(&trivial_map);
+            }
+
+            type E<C, Col> = Expr<C, ColumnExtended<Col>>;
+            let e = match self {
+                e @ Expr::VanishesOnLast4Rows
+                | e @ Expr::UnnormalizedLagrangeBasis(_)
+                | e @ Expr::Constant(_)
+                | e @ Expr::Cell(_) => e.map_column(&trivial_map),
+
+                Expr::Double(e) => E::<C, Col>::Double(Box::new(e.reduce_degree_to_2_rec(rec))),
+                Expr::Square(e) => E::<C, Col>::Square(Box::new(e.reduce_degree_to_1(rec))),
+                Expr::BinOp(op, e1, e2) => match op {
+                    op @ Op2::Add | op @ Op2::Sub => {
+                        let e1 = Box::new(e1.reduce_degree_to_2_rec(rec));
+                        let e2 = Box::new(e2.reduce_degree_to_2_rec(rec));
+                        E::<C, Col>::BinOp(op, e1, e2)
+                    }
+                    Op2::Mul => {
+                        let e1 = Box::new(e1.reduce_degree_to_1(rec));
+                        let e2 = Box::new(e2.reduce_degree_to_1(rec));
+                        E::<C, Col>::BinOp(Op2::Mul, e1, e2)
+                    }
+                },
+                //only supporting 1..=8 for simplicity, but arbitrary exponent could be
```

Comment:
You mean because in Kimchi we didn't support constraints which are higher than degree 8, right?